File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        
        if(n == 1)  return nums[0];
        
        int prev0 = nums[0];
        int prev1 = max(nums[0], nums[1]);
        
        for(int i = 2; i < n; i++){
            int cur = max(prev1, prev0 + nums[i]);
            prev0 = prev1;
            prev1 = cur;
        }
        
        return prev1;
    }
};

File: 0198-house-robber/0198-house-robber.java
class Solution {
    public int rob(int[] nums) {
        int n = nums.length;
        
        if(n == 1)  return nums[0];
        if(n == 2)  return Math.max(nums[0], nums[1]);
        
        int t[] = new int[n];
        
        t[0] = nums[0];
        t[1] = Math.max(nums[0], nums[1]);
        
        for(int i = 2; i < n; i++){
            t[i] = Math.max(t[i - 1], t[i - 2] + nums[i]);
        }
        
        return t[n - 1];
    }
}